sdk/nodejs/remote/index.ts

```diff
@@ -0,0 +1,107 @@
+            //    and populate this proxy from that URN.
+            // 2. A URN was not provided - in this case we are going to remotely construct the resource,
+            //    get the URN from the newly constructed resource, then look it up and populate this
+            //    proxy from that URN.
+            if (!opts.urn) {
+                const p = getRemoteServer().construct(libraryPath, libraryName, name, inputs, opts);
+                const urn = p.then(r => <string>r.urn);
+                opts = pulumi.mergeOptions(opts, { urn });
+            }
+            const props = {
+                ...inputs,
+                ...outputs,
+            };
+            super(t, name, props, opts);
+        }
+}
+
+let remoteServer: RemoteServer | undefined;
+function getRemoteServer(): RemoteServer {
+    if (!remoteServer) {
+        remoteServer = new RemoteServer();
+    }
+    return remoteServer;
+}
+
+class RemoteServer {
+    private client: Promise<any>;
+    constructor() {
+        const subprocess = cp.fork(require.resolve("./server"));
+        // Ensure we can exit the current process without waiting on the VM server process to exit.
+        subprocess.disconnect(); // detach the IPC connection
+        subprocess.unref(); // do not track subprocess on our event loop
+        this.client = new Promise(r => setTimeout(r, 1000)).then(() => {
+            return new runtimeServiceProto.RuntimeClient(
+                '0.0.0.0:50051',
+                grpc.credentials.createInsecure()
+            );
+        });
+    }
+
+    public async construct(libraryPath: string, resource: string, name: string, args: any, opts?: any): Promise<any> {
+        // TODO: Replace this with a proper wait on the server having launched (or retry).
+        await new Promise(r => setTimeout(r, 1000));
+        const serializedArgs = await runtime.serializeProperties("construct-args", args);
+        const argsStruct = gstruct.Struct.fromJavaScript(serializedArgs);
+        const serializedOpts = await runtime.serializeProperties("construct-opts", opts);
+        const optsStruct = gstruct.Struct.fromJavaScript(serializedOpts);
+        const client = await this.client;
+        const constructRequest = new runtimeProto.ConstructRequest();
+        constructRequest.setLibrarypath(libraryPath);
+        constructRequest.setResource(resource);
+        constructRequest.setName(name);
+        constructRequest.setArgs(argsStruct);
+        constructRequest.setOpts(optsStruct);
+        const outsStruct = await new Promise<any>((resolve, reject) => {
+            client.construct(constructRequest, (err: Error, resp: any) => {
+                if (err) {
+                    reject(err);
+                } else {
+                    resolve(resp.getOuts());
+                }
+            });
+        });
+        const outs = await runtime.deserializeProperties(outsStruct);
```

Comment:
it looks like the only thing you use construct for (so far) is to get the urn.  should that be hte only thing returned?

Reply:
Maybe - but I think it's  slightly cleaner to return the outs for now.  Once we implement function invoke here it may be clearer what's needed here (the top level returned object won't always necessarily be something with a `urn`). But this is also a very local internal contract, so easy to modify as needed in the future.